components/remerge/src/storage/db.rs

```diff
@@ -299,9 +308,42 @@ impl RemergeDb {
         Ok(vc.apply(self.client_id.clone(), counter))
     }
 
+    /// Returns NoSuchRecord if, well, there's no such record.
+    fn get_existing_record(&self, rec: &NativeRecord) -> Result<LocalRecord> {
+        use crate::{
+            schema::desc::{Field, FieldType},
+            JsonValue,
+        };
+        let native = self.info.native_schema();
+        let fidx = native
+            .field_own_guid
+            .expect("FIXME: own_guid should be explicitly mandatory");
+        let field = &native.fields[fidx];
+        assert!(
+            matches::matches!(field.ty, FieldType::OwnGuid { .. }),
+            "Validation/parsing bug -- field_own_guid must point to an own_guid"
+        );
+        // Just treat missing and null the same.
+        let val = rec.get(&field.local_name).unwrap_or(&JsonValue::Null);
+        let guid = Field::validate_guid(&field.local_name, val)?;
+
+        self.get_local_by_id(guid.as_str())?
+            .ok_or_else(|| ErrorKind::NoSuchRecord(guid.into()).into())
+    }
+
     pub fn update_record(&self, record: &NativeRecord) -> Result<()> {
-        let (guid, record) = self.info.native_to_local(record, ToLocalReason::Update)?;
         let tx = self.db.unchecked_transaction()?;
+
+        // fails with NoSuchRecord if the record doesn't exist.
+
+        // Potential optimization: we could skip this for schemas that don't use
+        // types which need `prev` (untyped_map, record_set, ...)
+        let prev = self.get_existing_record(&record)?;
```

Comment:
> That seems like a pretty big departure in terms of API.

This was more or less covered in slack, I think you said you were gonna add a PR with some new notes to the RFC to cement understanding.

> Hypothetically, this feels like it should be less safe in the face of concurrency

We have a transaction open during updates, so it should be fine, no?

> Maybe I would feel better if we had a docstring or something explaining that this might compute the update based on the current state of the database?

K

Reply:
I filed https://github.com/mozilla/application-services/pull/2470 to reflect my understanding of what we talked about on Slack. I don't have a problem with the design decision, just wanted to make sure I understood its rationale.